src/metafiles/binstall_v1.rs

```diff
@@ -0,0 +1,324 @@
+use std::{
+    borrow, cmp,
+    collections::{btree_set, BTreeSet},
+    fs, hash,
+    io::{self, Seek, Write},
+    iter::{IntoIterator, Iterator},
+    path::{Path, PathBuf},
+};
+
+use compact_str::CompactString;
+use miette::Diagnostic;
+use semver::Version;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use url::Url;
+
+use crate::{cargo_home, cratesio_url, create_if_not_exist, FileLock};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct MetaData {
+    pub name: CompactString,
+    pub version_req: CompactString,
+    pub current_version: Version,
+    pub source: Source,
+    pub target: CompactString,
+    pub bins: Vec<CompactString>,
+
+    /// Forwards compatibility. Unknown keys from future versions
+    /// will be stored here and retained when the file is saved.
+    ///
+    /// We use an `Vec` here since it is never accessed in Rust.
+    #[serde(flatten, with = "tuple_vec_map")]
+    pub other: Vec<(CompactString, serde_json::Value)>,
+}
+
+impl borrow::Borrow<str> for MetaData {
+    fn borrow(&self) -> &str {
+        &self.name
+    }
+}
+
+impl PartialEq for MetaData {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name
+    }
+}
+impl Eq for MetaData {}
+
+impl PartialOrd for MetaData {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        self.name.partial_cmp(&other.name)
+    }
+}
+
+impl Ord for MetaData {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        self.name.cmp(&other.name)
+    }
+}
+
+impl hash::Hash for MetaData {
+    fn hash<H>(&self, state: &mut H)
+    where
+        H: hash::Hasher,
+    {
+        self.name.hash(state)
+    }
+}
+
+#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
+pub enum SourceType {
+    Git,
+    Path,
+    Registry,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Source {
+    pub source_type: SourceType,
+    pub url: Url,
+}
```

Comment:
This could just be the enum used in cvs but with the right serde `tag` attribute applied: `#[serde(tag = "source_type")]`

Reply:
Somehow it failed with this:

```
thread 'metafiles::binstall_v1::test::rw_test' panicked at 'called `Result::unwrap()` on an `Err` value: S
erdeJsonParse(Error("cannot serialize tagged newtype variant Source::Registry containing a string", line:
0, column: 0))', src/metafiles/binstall_v1.rs:274:53
```

I think I will stick to the existing `Source`.